OBJC_INTEROP.md

```diff
@@ -0,0 +1,174 @@
+as extensions. That's why these declarations can't be overridden in Kotlin.
+And extension initializers aren't available as Kotlin constructors.
+
+### NSNumber
+
+While Kotlin primitive types in some cases are mapped to `NSNumber`
+(e.g. when they are boxed), `NSNumber` type is not automatically translated 
+to Kotlin primitive types when used as Swift/Objective-C parameter type or return value.
+The reason is that `NSNumber` type doesn't provide enough information
+about wrapped primitive value type, i.e. `NSNumber` is statically not known
+to be e.g. `Byte`, `Boolean` or `Double`. So Kotlin primitive values 
+should be cast to/from `NSNumber` manually (see [below](#casting-between-mapped-types)).
+
+### NSMutableString
+
+`NSMutableString` Objective-C class is not available from Kotlin.
+All instances of `NSMutableString` are copied when passed to Kotlin.
+
+### Collections
+
+Kotlin collections are converted to Swift/Objective-C collections as described
+by the table above. Swift/Objective-C collections are mapped to Kotlin in the same way,
+except for `NSMutableSet` and `NSMutableDictionary`. `NSMutableSet` isn't converted to
+Kotlin `MutableSet`. To pass an object for Kotlin `MutableSet`,
+one can create this kind of Kotlin collection explicitly by either creating it 
+in Kotlin with e.g. `mutableSetOf()`, or using `${prefix}MutableSet` class in
+Swift/Objective-C, where `prefix` is the framework names prefix.
+The same holds for `MutableMap`.
+
+### Function types
+
+Kotlin function-typed objects (e.g. lambdas) are converted to 
+Swift functions / Objective-C blocks. However there is a difference in how
+types of parameters and return values are mapped when translating a function
+and a function type. In the latter case primitive types are mapped to their
+boxed representation, `NSNumber`. Kotlin `Unit` return value is represented
+as corresponding `Unit` singleton in Swift/Objective-C. The value of this singleton
+can be retrieved in the same way as for any other Kotlin `object`
+(see singletons in the table above).
+To sum the things up:
+```
+fun foo(block: (Int) -> Unit) { ... }
+```
+
+would be represented in Swift as
+
+```
+func foo(block: (NSNumber) -> KotlinUnit)
+```
+
+and can be called like
+```
+foo {
+    bar($0 as! Int32)
+    return KotlinUnit()
+}
+```
+
+## Casting between mapped types
+
+When writing Kotlin code, an object may require to be converted from Kotlin type
+to equivalent Swift/Objective-C type (or vice versa). In this case plain old
+Kotlin cast can be used, e.g.
+```
+val nsArray = listOf(1, 2, 3) as NSArray
+val string = nsString as String
+val nsNumber = 42 as NSNumber
+```
+
+## Subclassing
+
+Kotlin classes and interfaces can be subclassed by Swift/Objective-C classes
+and protocols.
+Currently a class that adopts Kotlin protocol should inherit `NSObject`
+(either directly or indirectly). Note that all Kotlin classes do inherit `NSObject`,
+so a subclass of Kotlin class can adopt Kotlin protocol.
```

Comment:
Not entirely clear to me. Maybe "subclass of Kotlin class" -> "Objective-C subclass of Kotlin class"?

Reply:
It makes sense.